### PR TITLE
Add [RequiresPreviewFeaturesAttribute] to Http3

### LINKED ISSUE #34738: Indicate preview status for HTTP/3 in .NET 6
Place preview attribute on `HttpProtocols` enum for HTTP/3 values.

https://github.com/dotnet/aspnetcore/blob/04c7c65cef14d82dbb8cef60b2a8688b051f45b5/src/Servers/Kestrel/Core/src/HttpProtocols.cs#L18-L19

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Runtime/Versioning/RequiresPreviewFeaturesAttribute.cs

### PR DESCRIPTION
**Pull request by wtgodbe**
Fixes https://github.com/dotnet/aspnetcore/issues/34738. Confirmed in ILSpy that the attribute is present on the 2 enum values.

### CODE REVIEW

**Comment by Tratcher** [comment]
How does this show up in the sample app, as a warning?
https://github.com/dotnet/aspnetcore/blob/eb0f414830a5468d0fba0c885b0a9ed0227292f0/src/Servers/Kestrel/samples/Http3SampleApp/Program.cs#L48
Do we need a suppression there?

**Comment by wtgodbe** [comment]
> How does this show up in the sample app, as a warning?

> Unhandled exception. System.NotSupportedException: QUIC is not supported or enabled on this platform. See https://aka.ms/aspnet/kestrel/http3reqs for details.

So I guess we either need to suppress it, or have the sample require preview features as well?

**Comment by Tratcher** [comment]
That's a runtime error about your machine not supporting quic. The preview attribute should have given you a compile time warning/error?

**Comment by wtgodbe** [comment]
Nope, no warnings/errors on building the app

**Comment by Tratcher** [comment]
Hmm, then it's not working and we need to figure out why before merging it.

**Comment by wtgodbe** [comment]
@pgovind @jeffhandley @jkotas this PR adds the `RequiresPreviewFeaturesAttribute` to the enum values for the Http3 protocol, yet our sample using that protocol gives no warning: https://github.com/dotnet/aspnetcore/blob/eb0f414830a5468d0fba0c885b0a9ed0227292f0/src/Servers/Kestrel/samples/Http3SampleApp/Program.cs#L48 This is surprising, are we doing something wrong?

**Comment by jeffhandley** [comment]
The analyzer is currently set to produce `Info` level diagnostics. We are still refining the implementation ahead of increasing it to Warning/Error. Can you check to see if there was an `Info` message produced? If you are getting those, then you could proceed in putting your suppression in ahead of us increasing the level.

https://github.com/dotnet/roslyn-analyzers/blob/release/6.0.1xx/src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md#:~:text=ca2252

**Comment by pgovind** [comment]
> Can you check to see if there was an Info message produced? If you are getting those, then you could proceed in putting your suppression in ahead of us increasing the level.

Either this, or you can change the severity of the analyzer from `Info` to `Error` and see if VS reports errors. I'm almost certain it will. We have unit tests testing this exact scenario in the analyzer.

**Comment by wtgodbe** [comment]
Yeah, I do get `Using 'Http3' requires opting into preview features. See https://aka.ms/dotnet-warnings/preview-features for more information.` Adding a NoWarn suppresses it.

**Comment by JamesNK** [comment]
https://aka.ms/dotnet-warnings/preview-features links to Bing?

@pgovind @jeffhandley @jkotas Is there a reason there is no way to add a custom message with this attribute?

**Comment by jeffhandley** [comment]
We typically don't set up the aka.ms links until documentation is in place and published, which won't happen until RC1 ships.

I don't recall if the API Design Review discussion included potentially having a custom message. @pgovind / @terrajobst, was there a good reason not to?

**Comment by pgovind** [comment]
> Is there a reason there is no way to add a custom message with this attribute?

Interesting. We hadn't considered that because we didn't have a use case for it. @terrajobst should decide if we want to enable it. I'll just point out that adding a message _might_ mean changes in S.P.CoreLib, so we should decide soon if we want it for .NET6 (or we just add it as a new features for .NET 7)

**Comment by JamesNK** [comment]
For the context of HTTP/3, it would be nice to link to a doc about why the feature is preview, and how to manage using it, but it's not critical.

In .NET 7 you could add a constructor overload that takes a message for people to provide more detail.

**Comment by JamesNK** [comment]
Created issue: https://github.com/dotnet/runtime/issues/56498

File: src/Servers/Kestrel/Core/src/HttpProtocols.cs
Comment (terrajobst) [inline src/Servers/Kestrel/Core/src/HttpProtocols.cs:19]: Nit: one usually omits the suffix `Attribute` when using attributes

```suggestion
        [RequiresPreviewFeatures]
```

File: src/Servers/Kestrel/Core/src/HttpProtocols.cs
Comment (terrajobst) [inline src/Servers/Kestrel/Core/src/HttpProtocols.cs:21]:
```suggestion
        [RequiresPreviewFeatures]
```

**Review by terrajobst — CHANGES_REQUESTED** [review]
I believe this was briefly discussed a few days ago. Yes, I believe the usage of `[RequiresPreviewFeatures]` is appropriate here and would communicate our intent (the capability to use Http3 is there, but it's not supported and only provided as a preview).

**Review by terrajobst — APPROVED** [review]
Clicked the wrong button because I'm a PM

**Review by JamesNK — APPROVED** [review]

**Review by Tratcher — APPROVED** [review]

**Comment by terrajobst** [comment]
If we do this, is there a reason why we wouldn't mark this API?

```C#
namespace System.Net
{
    public static class HttpVersion
    {
        [RequiresPreviewFeatures]
        public static readonly Version Version30;
    }
}
````

@dotnet/ncl

**Comment by JamesNK** [comment]
A client can upgrade to use HTTP/3 without it being configured using that setting.

For example, Version = 2.0 and RequestedVersionOrHigher will upgrade to HTTP/3 without someone seeing the preview warning. That's why an appcontext switch is better for their scenario.

File: src/Servers/Kestrel/samples/Http3SampleApp/Http3SampleApp.csproj
Comment (adityamandaleeka) [inline src/Servers/Kestrel/samples/Http3SampleApp/Http3SampleApp.csproj:5]: Nit: might be good to add something like "because HTTP/3 is currently in preview state" or something here, so the intent is clear to someone in the future.

**Review by adityamandaleeka — APPROVED** [review]
LGTM aside from the minor nit.

File: src/Servers/Kestrel/samples/Http3SampleApp/Http3SampleApp.csproj
Comment (terrajobst) [inline src/Servers/Kestrel/samples/Http3SampleApp/Http3SampleApp.csproj:6]: Since this is a sample app, it should probably enable preview features, rather than disabling the warning, [see the design document](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md#enabling-a-preview-feature).

```suggestion
    <!-- Turn on preview features so we can use Http3 -->
    <EnablePreviewFeatures>True</EnablePreviewFeatures>
```

**Pull request merged**
